### PR TITLE
Spatial_Engine: Migrate Methods from Common_Engine

### LINKED ISSUE #1323: Common_Engine: IElementCurves uses BHoMObject
#### Description:
image
`IElement` is not a `BHoMObject`. This still "works" as it's mostly used for Tex `Bar` & `Panel` which also is a `BHoMObject`.

The only thing this does is sending it to a decrepit method which then casts it to `IElement` to then call the new one. Hence simply changing `BHoMObject` to `IElement` would fix this.

https://github.com/BHoM/BHoM_Engine/blob/977443201f405a70635299ab61c40f8a51827ebe/Common_Engine/Query/ElementCurves.cs#L86

### LINKED ISSUE #1470: Spatial_Engine: Migrate methods from Common_Engine
#### Suggestions to restore compliance:

Move the IElementXD methods from the Common_Engine to the Spatial_Engine .
As discussed with @al-fisher with the plan to finish the implementation by end of next sprint.
FYI @IsakNaslundBh @FraserGreenroyd 

after #1469 

With the restructuring of where the Interface methods sit in the Engine, as to add clarity.
i.e. If one implements one of `IElement` these are the method that BHoM expect the class to have an extension method for.

##### IElement0D
``` C#
IGeometry(this IElement0D)
 
ISetGeometry(this IElement0D, Point)
```
##### IElement1D
``` C#
IGeometry(this IElement1D)
 
IElements0D(this IElement1D)
INewElement0D(this IElement1D, Point)
 
ISetElements0D(this IElement1D, List<IElement0D>)
 
ISetGeometry(this IElement1D, ICurve)
```
##### IElement2D
``` C#
IOutlineElements1D(this IElement2D)
IInternalElements2D(this IElement2D)
 
INewElement1D(this IElement2D, ICurve)
INewInternalElement2D(this IElement2D)
 
ISetOutlineElements1D(this IElement2D, List<IElement1D>)
ISetInternalElements2D(this IElement2D, List<IElement2D>)
```

### PR DESCRIPTION
**Pull request by kThorsager**
### Issues addressed by this PR

Closes #1470 
Closes #1323 
The methods which were in the Common_Engine and using `IElement` have been moved to the Spatial_Engine.

The methods in the Common_Engine now calls the new methods in the Spatial_Engine and are set as Deprecated

#### The Common_Engine methods
https://burohappold.sharepoint.com/:f:/s/BHoM/EkvCJbx7B_FMgCy56DDGfYEBxAYsn8yH10yFp5eqifX0JQ?e=R83f7s
https://burohappold.sharepoint.com/:u:/s/BHoM/Ec_mnMQcCUhGnsKUvm4QmQABsvvFgHg2CxXnVu98VQ5CIw?e=NPnW9I
#### The Spatial_Engine methods
https://burohappold.sharepoint.com/:f:/s/BHoM/EjAN0bkxublBirJzGJFTg8YBo5ZncAd1-AKSzDYJBPRoGQ?e=rKz1dB
https://burohappold.sharepoint.com/:u:/s/BHoM/EZpL86wOI89CgPhZhFXOROABs6rsYFynGUYKPn9Uchc4Yg?e=suXMbb
#### ElementCurves before/after
note that on this PR this will always work as the Common_Engine method calls the Spatial_Engine method
https://burohappold.sharepoint.com/:f:/s/BHoM/Epamw8TSDxxHjlHkazmmyHoBAjVUpiJT-6HYBUz_443CAQ?e=P6QeHE

### Changelog

- `IElementCurves` is now usable with non-`BHoMObject`s

### Additional comments
Considering whether to add the descriptions etc right away or handle in separate PR.
Will the compliance check allow that?
@FraserGreenroyd  @al-fisher @pawelbaran

### CODE REVIEW

**Comment by kThorsager** [comment]
I'm unaware if Versioning looks at the Deprecated attributes (other than to attempt to Version) and since neither the name nor the class have changed i figured it redundant.

and I'm also unsure of what exactly the `Type newType` refers to (output, input, which input?)

But if that's needed I'll change that.

Also unsure if the Versioning likes needing to first version from `Geomerty.IElement` => `Dimensional.IElement` and then `Engine.Common` => `Engine.Spatial` as the chaining of versioning is per quarter. (we could perhaps set the versioning to go directly from `Geometry.IElement` => `Engine.Spatial` but that would miss all scripts done in the meantime) 

correct me if I'm wrong @adecler

**Comment by pawelbaran** [comment]
Right, both class and method names stay the same - thanks for answering, I will review the code today.

**Comment by IsakNaslundBh** [comment]
> I'm unaware if Versioning looks at the Deprecated attributes (other than to attempt to Version) and since neither the name nor the class have changed i figured it redundant.
> 
> and I'm also unsure of what exactly the `Type newType` refers to (output, input, which input?)
> 
> But if that's needed I'll change that.
> 
> Also unsure if the Versioning likes needing to first version from `Geomerty.IElement` => `Dimensional.IElement` and then `Engine.Common` => `Engine.Spatial` as the chaining of versioning is per quarter. (we could perhaps set the versioning to go directly from `Geometry.IElement` => `Engine.Spatial` but that would miss all scripts done in the meantime)
> 
> correct me if I'm wrong @adecler

Maybe we can add upgrade for both? but might be to painful. This is, and Method upgrade in versioning toolkit from common to spatial for all methods moved, both using IElements from the new position, as well as the old, if that makes sense?

**Comment by kThorsager** [comment]
> Maybe we can add upgrade for both? but might be to painful. This is, and Method upgrade in versioning toolkit from common to spatial for all methods moved, both using IElements from the new position, as well as the old, if that makes sense?

Maybe, but that would preferably be done with a method namespace converter, not by explicitly stating every method which has been moved and their new location.

I think this would be possible with changes to the Versioning_Toolkit around here:
image
to do some more sophisticated parsing of the type string beforehand so that the TypeConverter doesn't need to specify Version etc (as I'm unsure if that really is the same for all of the methods)

(This is where it checks if there's a equivalent string in the `TypeConverter.cs` `ToNewType` Dictionary)

**Comment by FraserGreenroyd** [comment]
I've taken a look, I'm also not keen on the `_Interface` folder for storing the methods - engines should stick with the standard folders - I see no planning or reasoning behind a sudden change on this one I'm afraid? 😸

**Comment by kThorsager** [comment]
The _Interface folder is added as they contain the methods which are called through reflection to the objects implementing `IElement` and do as a result stand apart as the base methods for interaction between the interface `IElement` and its objects. Mainly to convey to those looking to implement their object as a `IElement` can find what's needed easily.

I considered the `BHoMObject` to `IElement` fine since it kept the method from assuming that `IElement` was a `BHoMObject` and all subsequent methods was for `IElement` (and I've had confirmation that it was a mistake)
image
But a testfile is easy enough so I'll add that.

And I guess I considered a migration to not need testfiles due to no methods being removed or (to any large part) changed, but this is a good a time as any to add Common_Engines standard testfiles so I'll get on that

**Comment by pawelbaran** [comment]
> * As for the `_Interface` folder  itself. I actually get the need for this. @kThorsager I do think it is important for all the interface methods to be clearly distinguished. I might propose if you named all the files with there prefix of _"I"_ to match their interface method names. That will clearly identify them amongst the other methods? negating the need for a dedicated folder?

Not sure if we even need an _I_ prefix - so far we survived without it even with more core concepts than `IElements` (please correct me if I am wrong). But no strong opinion about it on my end.

**Comment by kThorsager** [comment]
Yes, I-prefix coupled with [the documentation page](https://github.com/BHoM/documentation/wiki/IElement-required-extension-methods) should keep things clear in a way which I did not feel when I proposed the separate folder.

**Comment by adecler** [comment]
> I've taken a look, I'm also not keen on the `_Interface` folder for storing the methods - engines should stick with the standard folders - I see no planning or reasoning behind a sudden change on this one I'm afraid? 😸

Fully agree here. Those interface methods are nothing more than an evolution of the interface methods using `as dynamic` as a dispatching solution. So it seems logical that they follow the same established conventions: At the top of the file in a `Interface Methods` section. 

I would also like to insist of the importance of not silently pushing new conventions that break the framework rules. Those need to be explicitly raised as a separate point of discussion. Thanks @pawelbaran and @FraserGreenroyd for raising your concerns. Otherwise I would not even noticed this was happening.

**Comment by al-fisher** [comment]
> Fully agree here. Those interface methods are nothing more than an evolution of the interface methods using as dynamic as a dispatching solution. So it seems logical that they follow the same established conventions: At the top of the file in a Interface Methods section.

👍👍

**Comment by al-fisher** [comment]
Reviewing the code now - looks good. Happy that all the concerns above by the team have been addressed.
Thanks @kThorsager 

Naturally there is the issue with respect to the naming convention and namespace for the so called "New" methods - this has been raised here: https://github.com/BHoM/BHoM_Engine/issues/1572 and should be closed out separately.

**Comment by kThorsager** [comment]
I got the Versioning to work for this if one removes the methods which I have now set a `Deprecated`, which should be fine, apart from the need to update every toolkit which uses the Common_Engine to Spatial_Engine this close to the beta release. 

I know that at least Revit and ETABS are affected, but have not done a closer query on it.

Not sure on how to get the versioning to work on it otherwise, so let me know if I should proceed?
@al-fisher @IsakNaslundBh @FraserGreenroyd @pawelbaran

**Comment by IsakNaslundBh** [comment]
Would not delete these methods now, as it requires more coordination across multiple repos as you say @kThorsager . Aim with this PR should be to make sure that all the new methods have been ported over correctly, and have tests to prove that is the case, and that all the to-be-deleted methods have been deprecated. Then in 3.2 we can calmly evaluate the impact of deleting them and make sure we raise PRs accordingly.

For the Versioning, as long as it is correct there is no real harm getting it merged in 3.1, as it only kicks in if it fails, but no real rush either (except for not keeping PRs open forever). Only important for it to be merged before we wipe all the common engine methods.

Priority now should be to make sure we have sufficient testfiles for the migrated methods.

**Comment by kThorsager** [comment]
I have now added test-files for most of the methods, is it needed for every method, because in that case I'll need some help in both understanding what `DistributeOutlines()` does and how I'm supposed to call it in the canvas.

Otherwise this is ready

**Comment by kThorsager** [comment]
> For the Versioning, as long as it is correct there is no real harm getting it merged in 3.1, as it only kicks in if it fails, but no real rush either (except for not keeping PRs open forever). Only important for it to be merged before we wipe all the common engine methods in 3.2.

Will give a problem when we switch over to Nugets in the Versioning_Toolkit and the BHoMUpgrader31 only can 'see' the code from the beta release, and hence never proc if we make the changes which it would react to later.

**Comment by IsakNaslundBh** [comment]
> > For the Versioning, as long as it is correct there is no real harm getting it merged in 3.1, as it only kicks in if it fails, but no real rush either (except for not keeping PRs open forever). Only important for it to be merged before we wipe all the common engine methods in 3.2.
> 
> Will give a problem when we switch over to Nugets in the Versioning_Toolkit and the BHoMUpgrader31 only can 'see' the code from the 3.1 beta release, and hence never proc if we make the changes which it would react to later.

The methods to be upgraded are referenced by string, so should not be any issue there?

We will need to be able to reference things no longer existing in the versioning toolkit (by string) to be able to delete things at all.

**Comment by kThorsager** [comment]
> > > For the Versioning, as long as it is correct there is no real harm getting it merged in 3.1, as it only kicks in if it fails, but no real rush either (except for not keeping PRs open forever). Only important for it to be merged before we wipe all the common engine methods in 3.2.
> > 
> > 
> > Will give a problem when we switch over to Nugets in the Versioning_Toolkit and the BHoMUpgrader31 only can 'see' the code from the 3.1 beta release, and hence never proc if we make the changes which it would react to later.
> 
> The methods to be upgraded are referenced by string, so should not be any issue there?
> 
> We will need to be able to reference things no longer existing in the versioning toolkit (by string) to be able to delete things at all.

Yes, you're right, I mixed up my thoughts
It's the methods which it upgrades into which must be in the same version,
sending it as the correct string happens on a UI level and should be fine

**Comment by kThorsager** [comment]
Checked the test file @pawelbaran and it looks fine for me, are you using Rhino5?
The missing component is Grasshoppers subtraction which was updated between 5 and 6

File: Spatial_Engine/Query/InternalElements2D.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace BH.Engine.Spatial
+{
+    public static partial class Query
+    {
+        /******************************************/
+        /****            IElement2D            ****/
+        /******************************************/
+
+        public static List<IElement2D> IInternalElements2D(this IElement2D element2D)
```
[the remark this answers is not in the document]
Comment (kThorsager) [reply Spatial_Engine/Query/InternalElements2D.cs]: No? @IsakNaslundBh

[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Spatial_Engine/Query/InternalElements2D.cs]: correct, ignore this

**Comment by pawelbaran** [comment]
You are right @kThorsager, I was testing on Rhino5 and I got a subtraction error. Makes sense.

**Comment by kThorsager** [comment]
All test-files are now in the folder on SharePoint, so this is ready for review

**Comment by IsakNaslundBh** [comment]
> All test-files are now in the folder on SharePoint, so this is ready for review

Well done @kThorsager . Will test things through!

**Pull request merged**